src/library_dylink.js

```diff
@@ -694,32 +686,24 @@ var LibraryDylink = {
       return;
     }
 
-    // if we can load dynamic libraries synchronously, do so, otherwise, preload
-    if (!readBinary) {
-      // we can't read binary data synchronously, so preload
-      addRunDependency('preloadDylibs');
-      dynamicLibraries.reduce(function(chain, lib) {
-        return chain.then(function() {
-          return loadDynamicLibrary(lib, {loadAsync: true, global: true, nodelete: true, allowUndefined: true});
-        });
-      }, Promise.resolve()).then(function() {
-        // we got them all, wonderful
-        removeRunDependency('preloadDylibs');
-        reportUndefinedSymbols();
+    // Load binaries asynchronously
+    addRunDependency('preloadDylibs');
+    dynamicLibraries.reduce(function(chain, lib) {
+      return chain.then(function() {
+        return loadDynamicLibrary(lib, {loadAsync: true, global: true, nodelete: true, allowUndefined: true});
       });
-      return;
-    }
-
-    dynamicLibraries.forEach(function(lib) {
-      // libraries linked to main never go away
-      loadDynamicLibrary(lib, {global: true, nodelete: true, allowUndefined: true});
+    }, Promise.resolve()).then(function() {
+      // we got them all, wonderful
+      reportUndefinedSymbols();
+      removeRunDependency('preloadDylibs');
+#if DYLINK_DEBUG
+    err('preloadDylibs done!');
+#endif
     });
-    reportUndefinedSymbols();
   },
 
   // void* dlopen(const char* filename, int flags);
   dlopen__deps: ['$DLFCN', '$FS', '$ENV'],
-  dlopen__proxy: 'sync',
```

Comment:
why are these changed here?

Reply:
Otherwise the dlopen will get proxied back to then the main thread.. where it can't block.

The list of loaded modules is worker specific so we don't want to be proxying dynamic loading calls back to the main thread.

Reply:
Makes sense, except don't we (eventually?) actually want a central authority to dlopen things, and communicate that to the other threads?

Reply:
Eventually we will need some kind of co-ordination to keep the threads in sync.   However, I imagine that the solution we use will still involve each thread having its own state (since each thread has its own wasmTable which could be out of sync with the other threeds).   Exactly how the threads are kept in sync of catch up with each other is still an open question.   

Also, it it seems like the thread that calls `dlopen` will probably want to load the dll in its own context right away.